pipelines/batch/src/main/java/org/openmrs/analytics/FetchResources.java

```diff
@@ -89,10 +93,33 @@ static String getSubjectPatientIdOrNull(Resource resource) {
 	
 	static class SearchFn extends FetchSearchPageFn<SearchSegmentDescriptor> {
 		
+		// This is to optimize counting number of patient resources. This will be initialized in
+		// StartBundle; we don't need to `synchronized` access of it because each instance of the DoFn
+		// class is accessed by a single thread.
+		// https://beam.apache.org/documentation/programming-guide/#user-code-thread-compatibility
+		private Map<String, Integer> patientCount = null;
+		
 		SearchFn(FhirEtlOptions options, String stageIdentifier) {
 			super(options, stageIdentifier);
 		}
 		
+		@StartBundle
+		public void startBundle(StartBundleContext context) {
+			patientCount = Maps.newHashMap();
+		}
+		
+		private void incrementPatientResources(String patientId) {
+			int current = patientCount.getOrDefault(patientId, 0);
+			patientCount.put(patientId, current + 1);
+		}
+		
+		@FinishBundle
+		public void finishBundle(FinishBundleContext context) {
+			for (Map.Entry<String, Integer> entry : patientCount.entrySet()) {
+				context.output(KV.of(entry.getKey(), entry.getValue()), Instant.now(), GlobalWindow.INSTANCE);
```

Comment:
Thank you for working on this. I am not sure how Global Window works. From the documentation  says using the default trigger

> The single global window with a default trigger generally requires the entire data set to be available before processing

Please do correct me if I'm wrong as I'm new to windowing, wouldn't this cause still cause the out of memory exception, since it will wait for all dataset to be available before processing?

I also see from the docs it recommend using non-default trigger 

>Set a non-default trigger. This allows the global window to emit results under other conditions, since the default windowing behavior (waiting for all data to arrive) will never occur.

Reply:
The fact that we need to specify a window here is a little bit confusing. Beam is a model that merges batch and stream processing together. So in this pipeline (which is batch) in fact we are only using the GlobalWindow (I know that we tried to benefit from some advantages of streaming/windowing before but we reverted that effort in PR #163). IOW, here and everywhere else, we expect that all processing might be done for one stage before the next one starts (so that is fine).

With that note, all we are doing in this PR is that instead of emitting a [patient, 1] pair for each resource and then group/count all of them, we group all patient IDs we see in a bundle of work, then emit a single [patient, COUNT] for each patient. So the memory saving here is that if a bundle of work has 10K resources in it but all those resources are for only 100 patients, then with this change we emit 100 patient-count pairs, instead of 10K.

I have reasons to believe that the large number of patient-count pairs at this stage is causing the memory issue you experienced but I have not proved that. Hence asking you to try it on the example you had.

Reply:
Thank you for the detailed explanation on the approach, makes sense to emit 100 patient pairs, instead of 10K.